### PR TITLE
[ADAM-1044] Support VCF annotation ANN field

### PR DESCRIPTION
**Pull request by heuermh**
Work towards #1044.
Will need further discussion around conversion from VariantContext as mentioned in #1062.

### CODE REVIEW

**Comment by heuermh** [comment]
Rebased.  This can go in as-is, the further work described above will go in #1078.

**Comment by fnothaft** [comment]
Thanks @heuermh! I will make a review pass today.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+package org.bdgenomics.adam.converters
+
+import htsjdk.variant.vcf.VCFConstants
+import htsjdk.variant.variantcontext.VariantContext
+import org.bdgenomics.formats.avro.{ TranscriptEffect, Variant, VariantAnnotation, VariantAnnotationMessage }
+import org.bdgenomics.formats.avro.VariantAnnotationMessage._
+import org.bdgenomics.utils.misc.Logging
+import scala.collection.JavaConverters._
+
+object VariantAnnotations extends Serializable with Logging {
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: VariantAnnotations --> VariantAnnotationsConverter, to be consistent with the rest of the package?

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bdgenomics.adam.converters
+
+import htsjdk.variant.vcf.VCFConstants
+import htsjdk.variant.variantcontext.VariantContext
+import org.bdgenomics.formats.avro.{ TranscriptEffect, Variant, VariantAnnotation, VariantAnnotationMessage }
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Split long line.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import org.bdgenomics.formats.avro.VariantAnnotationMessage._
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Prefer `._` imports above imports from same package.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+  private def notEmpty(s: String): Boolean = Option(s) match {
+    case None    => false
+    case Some(s) => s.length > 0
+  }
+
+  private[converters] def parseTranscriptEffect(s: String): Seq[TranscriptEffect] = {
+    val fields = s.split("\\|", -1)
+    if (fields.length < 16) {
+      if (log.isWarnEnabled) log.warn("Invalid VCF ANN attribute, expected at least 16 fields, found {}: {}", fields.length, fields)
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: What's the `log.isWarnEnabled` for? Might this be a better place to use a `ValidationStringency` check?

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    val tokens = s.split("/")
+    tokens.length match {
+      case 0 => (None, None)
+      case 1 => (Some(Integer.parseInt(tokens(0))), None)
+      case _ => (Some(Integer.parseInt(tokens(0))), Some(Integer.parseInt(tokens(1))))
+    }
+  }
+
+  private def notEmpty(s: String): Boolean = Option(s) match {
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Nit: prefer fold over match. E.g.:

```
Option(s).fold(false)(s => s.nonEmpty)
```

`nonEmpty` should come in implicitly from http://www.scala-lang.org/api/2.10.3/index.html#scala.collection.immutable.StringOps.

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Also, perhaps you could factor this a bit further into:

```
def setIfNotEmpty(s: String, setFn: String => Unit) {
  Option(s).filter(_.nonEmpty)
    .foreach(setFn)
}
```

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    // multivalued fields
+    val effects = parseEffects(fields(1))
+    val (rank, total) = parseFraction(fields(8))
+    val (cdnaPosition, cdnaLength) = parseFraction(fields(11))
+    val (cdsPosition, cdsLength) = parseFraction(fields(12))
+    val (proteinPosition, proteinLength) = parseFraction(fields(13))
+    val messages = parseMessages(fields(15))
+
+    val te = TranscriptEffect.newBuilder()
+    if (notEmpty(alternateAllele)) te.setAlternateAllele(alternateAllele)
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: If you do https://github.com/bigdatagenomics/adam/pull/1069/files#r71388866, then you can eliminate all of these if statements.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    total.foreach(te.setTotal(_))
+    if (notEmpty(transcriptHgvs)) te.setTranscriptHgvs(transcriptHgvs)
+    if (notEmpty(proteinHgvs)) te.setProteinHgvs(proteinHgvs)
+    cdnaPosition.foreach(te.setCdnaPosition(_))
+    cdnaLength.foreach(te.setCdnaLength(_))
+    cdsPosition.foreach(te.setCdsPosition(_))
+    cdsLength.foreach(te.setCdsLength(_))
+    proteinPosition.foreach(te.setProteinPosition(_))
+    proteinLength.foreach(te.setProteinLength(_))
+    if (notEmpty(distance)) te.setDistance(Integer.parseInt(distance))
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: I might try catch around this to throw a more informative error message if parsing fails.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    if (!messages.isEmpty) te.setMessages(messages.asJava)
+
+    Seq(te.build())
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Why are we returning a Seq here?

**Comment by fnothaft** [comment]
Generally looks good. Could you add scaladoc inline?

Also, where will this code attach in? Will it get connected up in #1078?

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala (hunk 1 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: In #1078 the code in this class will be merged with the already existing `VariantAnnotationConverter`.  Would you still want that change here?

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Oh, good point. Nevermind then!

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala (hunk 5 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Avro-generated classes track whether or not a field has been set (enabling compression for un-set fields I would presume) so it is important not to set a field where the value is null or empty.  This reads nice; for `null` the option won't be true, then `_.nonEmpty` won't NPE, and `setFn` is only called if `s` is non-empty.

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Exactly! I could wax philosophical about how I love the `Option` type in Scala...

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala (hunk 4 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: We don't have log level checks in `o.b.utils.Logging` methods, so this uses `log` directly.

`ValidationStringency` would be nice to have, but would need to be pushed down all the way through VCF parsing, and I'm not sure Hadoop-BAM supports it.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala (hunk 8 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: Same pattern used in feature parsing, `TranscriptEffect` association is an array in `VariantAnnotation` schema and a `java.util.List<TranscriptEffect>` in the Avro-generated object.

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala (hunk 4 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: > We don't have log level checks in o.b.utils.Logging methods

Really? Drat!

> `ValidationStringency` would be nice to have, but would need to be pushed down all the way through VCF parsing, and I'm not sure Hadoop-BAM supports it.

We call these functions in a `map` over the RDD of `VariantContext`s we get from Hadoop-BAM, so we can pass it down. I think it'd be good to have here, as it is likely desirable to fail on a badly formatted annotation line.

Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: ok, will try to push it down then

File: adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala (hunk 7 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/converters/VariantAnnotations.scala]: I've added the validation stringency stuff for invalid ANN lines but am not sure what to do for NumberFormatExceptions.  They aren't explicitly handled elsewhere.

**Comment by heuermh** [comment]
> Also, where will this code attach in? Will it get connected up in #1078?

Yes, I'll also add scaladoc at that point, since the method signatures will probably change to look more like the other converter classes.

**Comment by jpdna** [comment]
+1 - this looks good and seems self contained at least looking backwards, I am going to merge now so that related PRs can build on top of.

**Comment by jpdna** [comment]
Merged: c23aace7b173
Thanks @heuermh !

**Pull request closed without merging**